File: altgraph_web_app/Services/Graph/D3CsvBuilder.cs
namespace altgraph_web_app.Services.Graph
{

  public class D3CsvBuilder
  {
    public Graph? Graph { get; set; }
    public List<string> NodesCsvLines { get; set; } = new List<string>();
    public List<string> EdgeCsvLines { get; set; } = new List<string>();
    public Dictionary<string, string> CollectedNodesHash { get; set; } = new Dictionary<string, string>();
    public int CollectedNodesCount { get; set; }
    public Dictionary<string, string> CollectedEdgesHash { get; set; } = new Dictionary<string, string>();
    public int CollectedEdgesCount { get; set; }
    public string NodesCsvFile { get; set; } = "";
    public string EdgesCsvFile { get; set; } = "";
    private int iterationCount = 0;
    private readonly IConfiguration? _configuration;
    private readonly ILogger _logger;

    public D3CsvBuilder(Graph g, IConfiguration configuration, ILogger logger)
    {
      Graph = g;
      _configuration = configuration;
      _logger = logger;
      NodesCsvFile = _configuration["Paths:NodesCsvFile"];
      EdgesCsvFile = _configuration["Paths:EdgesCsvFile"];
    }

    public void BuildBillOfMaterialCsv(string sessionId, int depth)
    {
      CollectDataFromGraph(depth);
      BuildNodesCsv();
      BuildEdgesCsv();
      WriteCsvFilesAsync();
    }

    private void CollectDataFromGraph(int depth)
    {
      bool continueToCollect = true;
      string rootKey = Graph.RootKey;
      _logger.LogWarning($"collectLibrariesDataFromGraph rootKey: {rootKey}");

      string rootLib = ExtractNameFromKey(rootKey);
      _logger.LogWarning($"collectLibrariesDataFromGraph rootLib: {rootLib}");

      string rootEdgeKey = EdgeKey(rootLib, rootLib);
      _logger.LogWarning($"collectLibrariesDataFromGraph rootHashKey: {rootEdgeKey}");

      CollectedNodesHash[rootKey] = "pending";
      CollectedEdgesHash[rootEdgeKey] = "";

      while (continueToCollect)
      {
        iterationCount++;

        var currentKeys = CollectedNodesHash.Keys.ToList();
        for (int i = 0; i < currentKeys.Count; i++)
        {
          string currKey = currentKeys[i];
          string currVal = CollectedNodesHash[currKey];
          string currLib = ExtractNameFromKey(currKey);

          if (currVal.Equals("pending", StringComparison.OrdinalIgnoreCase))
          {
            GraphNode node = Graph.GraphMap[currKey];
            var dependencyKeys = node.AdjacentNodes.Keys.ToList();
            for (int d = 0; d < dependencyKeys.Count; d++)
            {
              string depKey = dependencyKeys[d];
              if (depKey.StartsWith("library"))
              {
                if (CollectedNodesHash.ContainsKey(depKey))
                {
                  _logger.LogWarning($"already in collectedNodesHash: {depKey}");
                }
                else
                {
                  CollectedNodesHash[depKey] = "pending";
                  string depLib = ExtractNameFromKey(depKey);
                  string depEdgeKey = EdgeKey(currLib, depLib);
                  CollectedEdgesHash[depEdgeKey] = "";
                }
              }
            }
            CollectedNodesHash[currKey] = "processed";
          }
        }

        // terminate the while-loop as necessary
        if (iterationCount == depth)
        {
          continueToCollect = false;
          _logger.LogError($"buildBillOfMaterialCsv while loop terminating at depth: {iterationCount}");
        }
        if (iterationCount >= 99)
        {
          continueToCollect = false;  // possible infinite loop, eject!
          _logger.LogError($"buildBillOfMaterialCsv while loop bailing out at iterationCount {iterationCount}");
        }
      }
    }

    private void BuildNodesCsv()
    {
      NodesCsvLines.Add("name,type,adjCount");  // header row
      List<string> keys = SortedArray(CollectedNodesHash.Keys.ToArray());
      for (int i = 0; i < keys.Count; i++)
      {
        string key = keys[i];
        string libName = ExtractNameFromKey(key);
        GraphNode node = Graph.GraphMap[key];
        NodesCsvLines.Add(libName + ",library," + node.AdjacentNodes.Count);
      }
      _logger.LogWarning($"buildNodesCsv count: {NodesCsvLines.Count}");
    }

    private void BuildEdgesCsv()
    {
      EdgeCsvLines.Add("source,target,weight");  // header row
      List<string> keys = SortedArray(CollectedEdgesHash.Keys.ToArray());
      for (int i = 0; i < keys.Count; i++)
      {
        string key = keys[i];
        string[] tokens = key.Split(" ");
        EdgeCsvLines.Add(tokens[0] + "," + tokens[1] + ",1");
      }
      _logger.LogWarning($"buildEdgesCsv count: {EdgeCsvLines.Count}");
    }

    private List<string> SortedArray(string[] array)
    {
      List<string> strings = new List<string>();
      for (int i = 0; i < array.Length; i++)
      {
        strings.Add((string)array[i]);
      }
      strings.Sort();
      return strings;
    }

    private string EdgeKey(string parentLibName, string childLibName)
    {
      return parentLibName + " " + childLibName;
    }

    private string ExtractNameFromKey(string key)
    {
      string name = key.Replace("^", ":").Split(":")[1].Replace(" ", "_");

      _logger.LogWarning($"extractNameFromKey: {key} -> {name}");
      // library^express^bf8cff83-5f7c-4995-8484-d2f405bcbce7^express -> express
      // author^TJ Holowaychuk <tj@vision-media.ca>^54dff427-35de-4a13-bcad-b3e4124b303a^TJ Holowaychuk <tj@vision-media.ca> -> TJ Holowaychuk <tj@vision-media.ca>

      return name;
    }

    private async void WriteCsvFilesAsync()
    {
      // GRAPH_NODES_CSV_FILE
      Directory.CreateDirectory(Path.GetDirectoryName(NodesCsvFile));
      Directory.CreateDirectory(Path.GetDirectoryName(EdgesCsvFile));

      await File.WriteAllLinesAsync(NodesCsvFile, NodesCsvLines);
      await File.WriteAllLinesAsync(EdgesCsvFile, EdgeCsvLines);
    }

    public void Finish()
    {
      Graph = null;
      CollectedNodesCount = CollectedNodesHash.Count();
      CollectedEdgesCount = CollectedEdgesHash.Count();
    }
  }

}

File: java/altgraph_data_app/src/main/java/org/cjoakim/cosmos/altgraph/data/common/graph/v1/D3CsvBuilder.java
package org.cjoakim.cosmos.altgraph.data.common.graph.v1;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.cjoakim.cosmos.altgraph.data.common.DataConstants;
import org.cjoakim.cosmos.altgraph.data.common.io.FileUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

/**
 * Instances of this class are created in the Web Application to create
 * the two necessary CSV files for D3.js to create a Graph visualization with,
 * from the given Graph object.  Used in the NPM graph.
 * <p>
 * Chris Joakim, Microsoft, November 2022
 */

@Data
@Slf4j
@JsonInclude(JsonInclude.Include.NON_NULL)
public class D3CsvBuilder implements DataConstants {

    private Graph graph;
    private ArrayList<String> nodesCsvLines = new ArrayList<String>();
    private ArrayList<String> edgeCsvLines = new ArrayList<String>();

    private HashMap<String, String> collectedNodesHash = new HashMap<String, String>();
    private int collectedNodesCount;
    private HashMap<String, String> collectedEdgesHash = new HashMap<String, String>();
    private int collectedEdgesCount;

    private String nodesCsvFile = null;
    private String edgesCsvFile = null;

    int iterationCount = 0;

    public D3CsvBuilder(Graph g) {
        super();
        graph = g;
        nodesCsvFile = NPM_GRAPH_NODES_CSV_FILE;
        edgesCsvFile = NPM_GRAPH_EDGES_CSV_FILE;
    }

    public void buildBillOfMaterialCsv(String sessionId, int depth) throws Exception {

        collectDataFromGraph(depth);
        buildNodesCsv();
        buildEdgesCsv();
        writeCsvFiles();
    }

    private void collectDataFromGraph(int depth) {

        boolean continueToCollect = true;
        String rootKey = graph.getRootKey();
        log.warn("collectLibrariesDataFromGraph rootKey: " + rootKey);

        String rootLib = extractNameFromKey(rootKey);
        log.warn("collectLibrariesDataFromGraph rootLib: " + rootLib);

        String rootEdgeKey = edgeKey(rootLib, rootLib);
        log.warn("collectLibrariesDataFromGraph rootHashKey: " + rootEdgeKey);

        collectedNodesHash.put(rootKey, "pending");
        collectedEdgesHash.put(rootEdgeKey, "");

        while (continueToCollect) {
            iterationCount++;
            Object[] currentKeys = collectedNodesHash.keySet().toArray();

            for (int i = 0; i < currentKeys.length; i++) {
                String currKey = (String) currentKeys[i];
                String currVal = collectedNodesHash.get(currKey);
                String currLib = extractNameFromKey(currKey);

                if (currVal.equalsIgnoreCase("pending")) {
                    GraphNode node = graph.getGraphMap().get(currKey);
                    Object[] dependencyKeys = node.getAdjacentNodes().keySet().toArray();
                    for (int d = 0; d < dependencyKeys.length; d++) {
                        String depKey = (String) dependencyKeys[d];
                        if (depKey.startsWith("library")) {
                            if (collectedNodesHash.containsKey(depKey)) {
                                //log.warn("already in collectedNodesHash: " + depKey);
                            } else {
                                collectedNodesHash.put(depKey, "pending");
                                String depLib = extractNameFromKey(depKey);
                                String depEdgeKey = edgeKey(currLib, depLib);
                                collectedEdgesHash.put(depEdgeKey, "");
                            }
                        }
                    }
                    collectedNodesHash.put(currKey, "processed");
                }
            }

            // terminate the while-loop as necessary
            if (iterationCount == depth) {
                continueToCollect = false;
                //log.error("buildBillOfMaterialCsv while loop terminating at depth: " + iterationCount);
            }
            if (iterationCount >= 99) {
                continueToCollect = false;  // possible infinite loop, eject!
                log.error("buildBillOfMaterialCsv while loop bailing out at iterationCount " + iterationCount);
            }
        }
    }

    private void buildNodesCsv() {
        nodesCsvLines.add("name,type,adjCount");  // header row
        ArrayList<String> keys = sortedArray(collectedNodesHash.keySet().toArray());
        for (int i = 0; i < keys.size(); i++) {
            String key = keys.get(i);
            String libName = extractNameFromKey(key);
            GraphNode node = graph.getGraphMap().get(key);
            nodesCsvLines.add(libName + ",library," + node.getAdjacentNodes().size());
        }
        log.warn("buildNodesCsv count: " + nodesCsvLines.size());
    }

    private void buildEdgesCsv() {
        edgeCsvLines.add("source,target,weight");  // header row
        ArrayList<String> keys = sortedArray(collectedEdgesHash.keySet().toArray());
        for (int i = 0; i < keys.size(); i++) {
            String key = keys.get(i);
            String[] tokens = key.split(" ");
            edgeCsvLines.add(tokens[0] + "," + tokens[1] + ",1");
        }
        log.warn("buildEdgesCsv count: " + edgeCsvLines.size());
    }

    private ArrayList<String> sortedArray(Object[] array) {

        ArrayList<String> strings = new ArrayList<>();
        for (int i = 0; i < array.length; i++) {
            strings.add((String) array[i]);
        }
        Collections.sort(strings);
        return strings;
    }

    private String edgeKey(String parentLibName, String childLibName) {

        return parentLibName + " " + childLibName;
    }

    private String extractNameFromKey(String key) {
        //String name = key.replace("^", " ").split(" ")[1];
        String name = key.replace("^", ":").split(":")[1].replace(" ", "_");

        // log.warn("extractNameFromKey: " + key + " -> " + name);
        // library^express^bf8cff83-5f7c-4995-8484-d2f405bcbce7^express -> express
        // author^TJ Holowaychuk <tj@vision-media.ca>^54dff427-35de-4a13-bcad-b3e4124b303a^TJ Holowaychuk <tj@vision-media.ca> -> TJ Holowaychuk <tj@vision-media.ca>

        return name;
    }

    private void writeCsvFiles() throws Exception {

        // GRAPH_NODES_CSV_FILE
        FileUtil fu = new FileUtil();
        fu.writeLines(nodesCsvFile, nodesCsvLines, true);
        fu.writeLines(edgesCsvFile, edgeCsvLines, true);
    }

    public String asJson(boolean pretty) throws Exception {
        try {
            ObjectMapper mapper = new ObjectMapper();
            if (pretty) {
                return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(this);
            } else {
                return mapper.writeValueAsString(this);
            }
        } catch (JsonProcessingException e) {
            return null;
        }
    }

    public void finish() {

        this.graph = null;
        collectedNodesCount = collectedNodesHash.size();
        collectedEdgesCount = collectedEdgesHash.size();
    }
}
